### PR TITLE
Limit number of connections per stream

### PR DESCRIPTION
**Pull request by mirpo**
Added number of connections for the stream. But we have InboundNodesForStream and OutboundNodesForStream, so now I check sum of both. But maybe we need separate limits? 
And also now we can get disconnection reason on protocol level.
@harbu with MAX_NUM_OF_NODES_PER_STREAM = 6 message-duplication.test.js is failing. So I removed 5th node. But maybe we want to add more tests for some edge cases.

### CODE REVIEW

File: src/protocol/NodeToNode.js [hunk 1]
```diff
@@ -42,6 +42,13 @@ class NodeToNode extends EventEmitter {
         this.endpoint.send(receiverNodeAddress, encoder.unsubscribeMessage(streamId))
     }
 
+    sendDisconnectionMessage(receiverNodeId, reason) {
```
Comment (harbu) [inline src/protocol/NodeToNode.js]: What about renaming to `disconnectFromNode` or `disconnect` since you are actually disconnecting (not just sending msg)?

File: src/logic/Node.js [hunk 2]
```diff
@@ -130,13 +132,18 @@ class Node extends EventEmitter {
 
         this.subscribeToStreamIfHaveNotYet(streamId)
 
-        this.streams.addOutboundNode(streamId, source)
-        if (!leechOnly) {
-            this.streams.addInboundNode(streamId, source)
+        if (this.streams.getInboundNodesForStream(streamId).length + this.streams.getOutboundNodesForStream(streamId).length >= MAX_NUM_OF_NODES_PER_STREAM) {
```
Comment (harbu) [inline src/logic/Node.js]: Btw each connected node will have an entry in both `inboundNodes` and `outboundNodes`, so now the limit becomes `MAX_NUM_OF_NODES_PER_STREAM / 2` WebSocket connections to nodes (per stream)

File: test/integration/message-duplication.test.js [hunk 3]
```diff
@@ -22,8 +22,7 @@ describe('duplicate message detection and avoidance', () => {
             startNetworkNode(LOCALHOST, 30352, 'node-1'),
             startNetworkNode(LOCALHOST, 30353, 'node-2'),
```
Comment (harbu) [inline test/integration/message-duplication.test.js]: Hmm.. I think this test should still work though? Because it is only checking upper limits not exact numbers. I'll try it out before change and see if I can see what the issue is.

**Review by harbu — CHANGES_REQUESTED** [review]
LGTM; few ideas + I think the test should still work so we should figure out what is the issue?

File: test/integration/message-duplication.test.js (hunk 3 above)
Comment (mirpo) [reply test/integration/message-duplication.test.js]: I put original test back, lets check together.

File: src/logic/Node.js [hunk 4]
```diff
@@ -132,18 +133,25 @@ class Node extends EventEmitter {
...
+            this.debug('reached max number "%d" for outbound connections for stream %s', MAX_NUM_NODES_OUTBOUND_PER_STREAM, streamId)
+            this.protocols.nodeToNode.disconnectFromNode(source, disconnectionReasons.MAX_OUTBOUND_CONNECTIONS)
+        }
+
+        if (!leechOnly) {
+            if (this.streams.getInboundNodesForStream(streamId).length < MAX_NUM_NODES_INBOUND_PER_STREAM) {
                 this.streams.addInboundNode(streamId, source)
+            } else {
+                this.debug('reached max number "%d" for inbound connections for stream %s', MAX_NUM_NODES_INBOUND_PER_STREAM, streamId)
+                this.protocols.nodeToNode.disconnectFromNode(source, disconnectionReasons.MAX_INBOUND_CONNECTIONS)
```
Comment (harbu) [inline src/logic/Node.js]: Method `nodeToNode.disconnectFromNode` may be invoked the 2nd time here if the earlier `else` block got executed?

File: src/logic/Node.js [hunk 5]
```diff
@@ -132,18 +133,25 @@ class Node extends EventEmitter {
...
             }
-            this._handleBufferedMessages(streamId)
-            this.debug('node %s subscribed to stream %s', source, streamId)
-            this.emit(events.SUBSCRIPTION_RECEIVED, streamId, source)
         }
+
+        this._handleBufferedMessages(streamId)
```
Comment (harbu) [inline src/logic/Node.js]: I think this code should get executed only if `disconnectFromNode` was not executed.

File: test/integration/connection-limits.test.js [hunk 6]
```diff
@@ -0,0 +1,57 @@
+const { startNetworkNode, startTracker } = require('../../src/composition')
+const { callbackToPromise } = require('../../src/util')
+const { LOCALHOST, DEFAULT_TIMEOUT } = require('../util')
+const endpointEvents = require('../../src/connection/Endpoint').events
+const { disconnectionReasons } = require('../../src/messages/messageTypes')
+
+jest.setTimeout(DEFAULT_TIMEOUT)
+
+/**
+ * This test verifies that on receiving a duplicate message, it is not re-emitted to the node's subscribers.
```
Comment (harbu) [inline test/integration/connection-limits.test.js:10]: This comment needs to be removed or re-written

File: test/integration/connection-limits.test.js [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+ */
+describe('duplicate message detection and avoidance', () => {
```
Comment (harbu) [inline test/integration/connection-limits.test.js:12]: Better description?

**Review by harbu — CHANGES_REQUESTED** [review]

File: src/logic/Node.js (hunk 4 above)
Comment (mirpo) [reply src/logic/Node.js]: Yes, but we check in WsEndpoint.close " if (!this.isConnected(recipientAddress))

**Review by ptesavol — APPROVED** [review]
Looks ok. Does the test go through now?
